src/main/java/com/minecolonies/coremod/colony/Structures.java

```diff
@@ -60,200 +109,845 @@ public static void init()
-                if (NULL_STYLE.equals(style) || MINER_STYLE.equals(style))
-                {
-                    continue;
-                }
-
                 if (path.toString().endsWith(SCHEMATIC_EXTENSION))
                 {
-                    final String filename = path.getFileName().toString().split("\\.nbt")[0];
-                    final String hut = filename.split("\\d+")[0];
+                    String relativePath = path.toString().substring(basePath.toString().length()).split("\\" + SCHEMATIC_EXTENSION)[0];
+                    if (!path.getFileSystem().getSeparator().equals("/"))
+                    {
+                        relativePath = relativePath.replace(path.getFileSystem().getSeparator(), "/");
+                    }
+                    if (relativePath.startsWith("/"))
+                    {
+                        relativePath = relativePath.substring(1);
+                    }
 
-                    if (isSchematicHut(hut))
+
+
+                    final StructureName structureName = new StructureName(relativePath);
+                    final String md5 = Structure.calculateMD5(Structure.getStream(relativePath));
+                    if (md5 == null)
                     {
-                        addHutStyle(hut, style);
-                        incrementHutMaxLevel(hut);
+                        Log.getLogger().error("Structures: " + structureName + " with md5 null.");
                     }
                     else
                     {
-                        addDecorationStyle(style, filename);
+
+                        if (md5Map.containsKey(structureName.toString()))
+                        {
+                            Log.getLogger().info("Override " + structureName + " md5:" + md5);
+                        }
+                        else
+                        {
+                            Log.getLogger().info("Add " + structureName + " md5:" + md5);
+                        }
+
+                        md5Map.put(structureName.toString(), md5);
+                        final int MAX_TOTAL_SIZE = 32767;
+                        final int MAX_SIZE = MAX_TOTAL_SIZE - Integer.SIZE / Byte.SIZE;
+                        final byte[] data = Structure.getStreamAsByteArray(Structure.getStream(structureName.toString()));
+                        final byte[] compressed = Structure.compress(data);
+
+                        if (compressed != null && compressed.length > MAX_SIZE)
+                        {
+                            Log.getLogger().warn("Structure " + structureName + " is " + compressed.length + " bytes when compress, maximum allower is " + MAX_SIZE + " bytes.");
+                        }
+
+                        if (MineColonies.isClient())
+                        {
+                            addSchematic(structureName);
+                        }
                     }
                 }
             }
         }
+        catch (@NotNull IOException e)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)